util: get clusterID for the passed in mon string

humblec opened the pull request:
as part of migration support, the clusterID has to be fetched
from passed in mon. Because the intree RBD storage class only
got monitor and not `clusterID` parameter support. However, in
CSI, SC has the `clusterID` parameter support but not mon. Due
to that we have to fetch the clusterID from config file for the
passed in mon and use it in our operations. This adds a helper
function to retrieve clusterID from passed in mon string.

Updates https://github.com/ceph/ceph-csi/issues/2509

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
util: get clusterID for the passed in mon string

as part of migration support, the clusterID has to be fetched
from passed in mon. Because the intree RBD storage class only
got monitor and not `clusterID` parameter support. However, in
CSI, SC has the `clusterID` parameter support but not mon. Due
to that we have to fetch the clusterID from config file for the
passed in mon and use it in our operations. This adds a helper
function to retrieve clusterID from passed in mon string.

Updates https://github.com/ceph/ceph-csi/issues/2509

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Madhu-1 reviewed (dismissed):
@humblec why do you need clusterID? without complete implementation, this PR is not complete.
can you please provide more details in the commit message and PR description?

humblec commented:
> @humblec why do you need clusterID? without complete implementation, this PR is not complete.
> can you please provide more details in the commit message and PR description?

sure, let me add details, its only in issue.. adding it

humblec commented:
> @humblec why do you need clusterID? without complete implementation, this PR is not complete.
> can you please provide more details in the commit message and PR description?

@Madhu-1 done.. ptal.  thanks ..

Madhu-1 reviewed (commented):
@humblec can you please complete the implementation instead of adding the helper function which is not going to get called anywhere in the code. and also if you have the mon IP then why do you need the clusterID and what you are going to do with clusterID?

Rakshith-R reviewed (dismissed):
@humblec , I think we can simulate the usage of this function/ online migration in our ci by using static rbd pv/pvc with such a volumeHandle (formed as described in the issue). 

By doing this we will be sure of the usage of util helper and the process of mounting, unmounting works. 

Can you please add such e2e test case[s].  
And add links to online migration docs and steps prescribed from kublet's side too.

edit; clusterID -> volumehandle

humblec committed:
rbd: fill clusterID if its a migration nodestage request

the migration nodestage request does not carry the 'clusterID' in it
and only monitors are available with the volumeContext. The volume
context flag 'migration=true' and 'static=true' flags allow us to
fill 'clusterID' from the passed in monitors to the volume Context,so
that rest of the static operations on nodestage can be proceeded as we
do treat static volumes today.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
> @humblec can you please complete the implementation instead of adding the helper function which is not going to get called anywhere in the code. and also if you have the mon IP then why do you need the clusterID and what you are going to do with clusterID?

@Madhu-1 ptal at the latest patchset, the top commit add nodestage request functionality for migrated volumes, the clusterID has to be retrieved from the monitors passed in and use it to get the volumes treated as static volume and rest of the static volume operations go through..

humblec commented:
> @humblec , I think we can simulate the usage of this function/ online migration in our ci by using static rbd pv/pvc with such a volumeHandle (formed as described in the issue).
> 
> By doing this we will be sure of the usage of util helper and the process of mounting, unmounting works.
> 
> Can you please add such e2e test case[s].
> And add links to online migration docs and steps prescribed from kublet's side too.
> 
> edit; clusterID -> volumehandle

@Rakshith-R ptal : so for migrated volumes, apart from today's static volume treatment, we just need to fill the clusterID from the available monitors, rest of the operations should be pretty straightforward.

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -264,6 +282,17 @@ func (ns *NodeServer) NodeStageVolume(
+	// Check this is a migration request because in that case, unlike other node stage requests
+	// it will be missing the clusterID, so fill it by fetching it from config file using mon.
+	if req.GetVolumeContext()["migration"] == "true" && req.VolumeContext["clusterID"] == "" {
+		cID, err := fillClusterIDFromMigrationReq(req.GetVolumeContext())
```
do error checking first and return if there are any errors.

humblec replied:
sure, done. thanks.

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -264,6 +282,17 @@ func (ns *NodeServer) NodeStageVolume(
+	if req.GetVolumeContext()["migration"] == "true" && req.VolumeContext["clusterID"] == "" {
+		cID, err := fillClusterIDFromMigrationReq(req.GetVolumeContext())
+		if cID != "" {
+			req.VolumeContext["clusterID"] = cID
```
i don't think it's a good approach to modify the existing request context. with this design looks like no other option to avoid minimal code changes.

Madhu-1 replied:
use predefined clusterID const from `csiConfig.go`

humblec replied:
sure, done..

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -188,6 +201,11 @@ func populateRbdVol(
 	// get rbd image name from the volume journal
 	// for static volumes, the image name is actually the volume ID itself
 	if isStaticVol {
+		if req.GetVolumeContext()["migration"] == "true" {
```
why we are adding both migration and staticVolume to the attributes? this is just to treat rbd images as static one even though its migrated?

humblec replied:
yep, an extra marker to make its a migrated static volume.

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -149,6 +149,19 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
+	rclusterID := ""
+	mons := parameters["monitors"]
+	for _, m := range strings.Split(mons, ",") {
+		rclusterID, err = util.GetClusterIDFromMon(util.CsiConfigFile, m)
```
please check for specific error messages (const) if the error is as expected then continue or else return the error.

humblec replied:
actually any error or iow, except `nil` for error is a signal to bail out , so avoiding an extra check for a specific error here.

Madhu-1 replied:
If there is an error in reading the configuration file when we need to continue in the for loop?

Rakshith-R replied:
```suggestion
		rclusterID, err = util.GetClusterIDFromMon(m)
```

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -188,6 +201,11 @@ func populateRbdVol(
+		if req.GetVolumeContext()["migration"] == "true" {
+			// if migration static volume, use imageName as volID
+			rv.RbdImageName = req.GetVolumeContext()["imageName"]
+			volID = rv.RbdImageName
```
why not send volID as the rbdImage name itself from the kubelet instead of doing a hack here? are you planning to add support for image deletion also?

humblec replied:
Yep, exactly , thats the plan to add deletion but only for migration static volume and not for other static volume (atleast in first version) :+1:

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -188,6 +201,11 @@ func populateRbdVol(
 	if isStaticVol {
+		if req.GetVolumeContext()["migration"] == "true" {
+			// if migration static volume, use imageName as volID
+			rv.RbdImageName = req.GetVolumeContext()["imageName"]
```
this looks redundant to me.

humblec replied:
the rv struct should have it , but kind of double confirmaion

Madhu-1 replied:
But do we need this double-check here? IMO it not needed

humblec replied:
imo, it looks like its not populated at this stage, I could be wrong, but making sure or assuring should not be of no harm as well.

Madhu-1 replied:
you are doing 
```
rv.RbdImageName = req.GetVolumeContext()["imageName"]
volID = rv.RbdImageName
```
inside the if check and again after if check there is code to assign `volID` to ` rv.RbdImageName`

this can be `volID=req.GetVolumeContext()["imageName"]` You don't need to assign it the imageName and volID and again assign volID to imageName

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -149,6 +149,19 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
 	return nil
 }
 
+func fillClusterIDFromMigrationReq(parameters map[string]string) (string, error) {
```
this can cause issues if the radosNamespace configuration is present for this clusterID.

Madhu-1 replied:
this comment is not addressed and the function name can be `GetClusterIDFromMigrationVolume`

humblec replied:
shouldnt be an issue in this case.

Madhu-1 replied:
@humblec why it's not an issue can you please explain?

humblec replied:
@Madhu-1 radosnamespace wont be passed in this migration case.

Madhu-1 replied:
even if you don't pass the radosnamespace will be taken from the existing configmap and https://github.com/ceph/ceph-csi/blob/devel/internal/rbd/rbd_util.go#L1178-L1181 this is going to get called and things will fail and if you have multiple clusterID with the same mons again not sure things will work as each clusterID's can/might contain different radosNamespaces

Madhu-1 replied:
@humblec is this not addressed yet, function name need to be changed

humblec replied:
Done.

Madhu-1 reviewed (dismissed):
@humblec I would like to see below things to make it complete

* Fix CI issue
* Add documentation for this one
* Are you planning to add the DeletionVolume option also?
* Add E2E test cases to cover this migration case with static PVC (krbd and nbd)

Madhu-1 commented:
@pkalever can you please take a look at it from an nbd point of view?

humblec committed:
e2e: Add e2e for rbd migration static pvc

This commit adds e2e for rbd migration static PVCs

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
> * Fix CI issue 
>
 Done
> * Add documentation for this one
>
Will cover it in a follow up PR , adding to todo tracker
> * Are you planning to add the DeletionVolume option also?:
>
Yes, will cover in a follow up PR 
> * Add E2E test cases to cover this migration case with static PVC (krbd and nbd)
>
 Done.

@Madhu-1 @Rakshith-R all comments are addressed , ptal.. Thanks.

humblec commented:
```
Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
```

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.20

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -189,6 +189,109 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
+	opt["pool"] = defaultRBDPool
+	opt["staticVolume"] = strconv.FormatBool(true)
+	opt["imageName"] = rbdImageName
+	if radosNamespace != "" {
+		opt["radosNamespace"] = radosNamespace
+	}
```
Does kubernetes support create rbd image with radosNamespace?

humblec replied:
dont see that  option, so will remove this.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -189,6 +189,109 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
+
+	opt["migration"] = "true"
+	opt["monitors"] = mon
+	if !checkImgFeat {
```
This value is always false do we need this one?

humblec replied:
just kept it for parity with other static test

Madhu-1 replied:
This code will never get hit then why do we need it?

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -189,6 +189,109 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
+	mon := "rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789"
+
+	size := "4Gi"
+	// create rbd image
+	cmd := fmt.Sprintf(
+		"rbd create %s --size=%d --image-feature=layering %s",
+		rbdImageName,
+		4096,
+		rbdOptions(defaultRBDPool))
```
kubernetes supports v1 and v2 images you might need to add tests for both?

humblec replied:
atleast not required now, can revisit later if really needed.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -189,6 +189,109 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
+	)
+
+	c := f.ClientSet
+	mon := "rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789"
```
please retrieve mon from the ceph cluster instead of hardcoding it.

humblec replied:
I actually did that yeterday but it didnt go through as mon was referred like this. regardless this is not a concern as its admin's responsibility to make sure the mapping exist. also unit tests are also in place to prove that things are working pretty good.

humblec replied:
regardless, also adding a TODO comment for later.

Rakshith-R replied:
https://github.com/ceph/ceph-csi/blob/22bb31df196ba72cb967a9f9d98e6e277315aa6f/e2e/utils.go#L69 

Mons can be fetched using this helper ?
(even though i think the service will lead to the same ip format each time.)

Madhu-1 replied:
This will fail if someone runs  with  different mons, suggest to use the already existing helper function

humblec replied:
yeah, will try to figure it out later as its not a priority one atleast for now.

Madhu-1 replied:
but why merge/push a code that will break if someone runs tests locally?

humblec replied:
@Madhu-1 addressed it.. thanks.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -189,6 +189,109 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
+	}
+	opt["pool"] = defaultRBDPool
+	opt["staticVolume"] = strconv.FormatBool(true)
+	opt["imageName"] = rbdImageName
```
this test cover krbd. kubernetes also supports nbd driver? are we not supporting it?

humblec replied:
considering both are in alpha, I will add nbd case as todo for now.

humblec replied:
Done., added in todo

Rakshith-R commented on internal/util/csiconfig.go:
```diff
@@ -154,10 +154,55 @@ func GetMonsAndClusterID(ctx context.Context, clusterID string, checkClusterIDMa
+// clusterID based on the passed in mon string. If passed in 'mon'
+// string has been found in the config the clusterID is returned,
+// else error.
+func GetClusterIDFromMon(pathToConfig, mon string) (string, error) {
+	clusterID, err := readClusterInfoWithMon(pathToConfig, mon)
```
```suggestion
func GetClusterIDFromMon(mon string) (string, error) {
	clusterID, err := readClusterInfoWithMon(CsiConfigFile, mon)
```

humblec replied:
done..

Rakshith-R commented on internal/rbd/nodeserver.go:
```diff
@@ -149,6 +149,21 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
 	return nil
 }
 
+// fillClusterIDFromMigrationReq fills the clusterID for the passed in monitors.
```
Move this function to util, as it not rbd specific ?

humblec replied:
the migration is only applicable for rbd now, so its fine to keep it here.

Rakshith-R commented on internal/util/csiconfig.go:
```diff
@@ -154,10 +154,55 @@ func GetMonsAndClusterID(ctx context.Context, clusterID string, checkClusterIDMa
+		return "", err
+	}
+
+	if clusterID == "" {
```
move this check to `readClusterInfoWithMon`, 

that way unit testcases for `readClusterInfoWithMon` will be sufficient.

humblec replied:
Done

pkalever commented on e2e/staticpvc.go:
```diff
@@ -189,6 +189,109 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
 	return err
 }
 
+func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBlock, checkImgFeat bool) error {
```
@humblec can we also consider improving this to take `mounter-type` parameter and act accordingly?
if so, it will be fairly simple to call this with krbd and rbd-nbd mounters from e2e/rbd.go.

Thanks!

humblec replied:
@pkalever yeah, can be done, will keep it for todo. thanks :+1:

pkalever replied:
[Just repeating here for context]
If the tests are not part of the same #PR, then probably it means we cannot validate `get clusterID` for the NBD case.
if that is fine with others, then we can create an issue (or add to some tracker) for persistence (set bot rule to keep-alive) and move on.

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -149,6 +149,27 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
+	for _, m := range strings.Split(mons, ",") {
+		rclusterID, err = util.GetClusterIDFromMon(m)
+		if err != nil {
+			if (strings.Contains(err.Error(), util.ErrConfFileFetchFailed)) ||
+				(strings.Contains(err.Error(), util.ErrUnmarshalFailed)) {
+				return "", err
```
you can check ` "missing configuration of cluster ID for mon %q", mon`(deing const in errors.go) and return error 
```suggestion
if err != nil && !errors.Is(err, util.ErrorMissingConfiguration){
return "",err
}
```

humblec replied:
Done.. @Madhu-1

Rakshith-R reviewed (commented):
LGTM, I'll let madhu finish his review.

humblec commented:
> LGTM, I'll let madhu finish his review.

Thanks.. adding the last 2 comments from madhu. :+1:

humblec commented:
@Madhu-1 @Rakshith-R all  the comments are addressed, ptal. thanks!

Madhu-1 commented on internal/util/csiconfig_test.go:
```diff
@@ -34,7 +34,7 @@ func cleanupTestData() {
 }
 
-// TODO: make this function less complex.
+// nolint:gocyclo,cyclop // reduce complexity
```
Please bring back the ToDo which can be used for tracking the todos in project

Madhu-1 commented on internal/util/errors.go:
```diff
@@ -35,6 +35,8 @@ var (
 	// ErrClusterIDNotSet is returned when cluster id is not set.
 	ErrClusterIDNotSet = errors.New("clusterID must be set")
+	// ErrMissingConfigForMon is returned when clusterID is not found for the mon.
+	ErrMissingConfigForMon = errors.New("missing configuration of cluster ID for mon")
```
For monitor

Madhu-1 commented on internal/util/csiconfig_test.go:
```diff
@@ -132,4 +132,29 @@ func TestCSIConfig(t *testing.T) {
 	}
+
+	data = "[{\"clusterID\":\"" + clusterID2 + "\",\"monitors\":[\"mon1\",\"mon2\",\"mon3\"]}," +
+		"{\"clusterID\":\"" + clusterID1 + "\",\"monitors\":[\"mon4\",\"mon5\",\"mon6\"]}]"
```
Add a test case for radosnamespace

humblec replied:
done.

humblec commented:
@Madhu-1 @Rakshith-R pending or all the comments are addressed. ptal.

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

humblec commented:
@Madhu-1 @Rakshith-R can you please take a final look at this PR ?

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -189,6 +189,106 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
+
+	app.Namespace = namespace
+	app.Spec.Volumes[0].PersistentVolumeClaim.ClaimName = pvcName
+	if checkImgFeat {
+		err = createAppErr(f.ClientSet, app, deployTimeout, "missing required parameter imageFeatures")
```
this part of the code will never get hit as you are passing `checkImgFeat` as always false. please remove `checkImgFeat` argument from the function

humblec replied:
done. thanks..

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -72,6 +72,10 @@ const (
 	thickProvisionMetaData = "true"
 	thinProvisionMetaData  = "false"
+
+	// these are the migration label key and value for params in volume context.
```
` params in volume context` to `parameters in volume context`

humblec replied:
done.. thanks..

Madhu-1 commented on internal/util/csiconfig.go:
```diff
@@ -154,10 +154,54 @@ func GetMonsAndClusterID(ctx context.Context, clusterID string, checkClusterIDMa
+	}
+
+	for _, cluster := range config {
+		if cluster.RadosNamespace != "" {
```
add  a comment why we are continuing in this case.

humblec replied:
sure, done .. thanks..

Madhu-1 commented on internal/util/errors.go:
```diff
@@ -35,6 +35,8 @@ var (
 	// ErrClusterIDNotSet is returned when cluster id is not set.
 	ErrClusterIDNotSet = errors.New("clusterID must be set")
+	// ErrMissingConfigForMonitor is returned when clusterID is not found for the mon.
+	ErrMissingConfigForMonitor = errors.New("missing configuration of cluster ID for mon")
```
`missing configuration of cluster ID for mon` to `missing configuration of a cluster ID for monitor`

humblec replied:
Done.. thanks ..

Madhu-1 reviewed (commented):
LGTM. small nits

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -189,6 +189,102 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
 	return err
 }
 
+func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBlock, checkImgFeat bool) error {
```
`checkImgFeat` not used in this function, needs to be removed.

humblec replied:
:+1: , done.. thanks

Rakshith-R commented on internal/util/csiconfig.go:
```diff
@@ -154,10 +154,57 @@ func GetMonsAndClusterID(ctx context.Context, clusterID string, checkClusterIDMa
+// else error.
+func GetClusterIDFromMon(mon string) (string, error) {
+	clusterID, err := readClusterInfoWithMon(CsiConfigFile, mon)
+	if err != nil {
+		return "", err
+	}
+
+	return clusterID, nil
```
```suggestion

	return clusterID, err
```

humblec replied:
Done.

humblec replied:
@Rakshith-R thanks.

Rakshith-R reviewed (dismissed):
@humblec , it looks good to me
just a single small change to remove redundant check.

humblec commented:
@Mergifyio rebase
